editor/components/post-permalink/index.js

```diff
@@ -14,55 +14,151 @@ import { Dashicon, ClipboardButton, Button } from '@wordpress/components';
+			editedPostName: '',
+			editingPermalink: false,
 		};
-		this.onCopy = this.onCopy.bind( this );
-		this.onFinishCopy = this.onFinishCopy.bind( this );
+
+		this.onSavePermalink = this.onSavePermalink.bind( this );
 	}
 
-	componentWillUnmount() {
-		clearTimeout( this.dismissCopyConfirmation );
+	componentWillMount() {
+		if ( this.props.samplePermalinkData ) {
+			this.updateSamplePermalink( this.props.samplePermalinkData[ 0 ], this.props.samplePermalinkData[ 1 ] );
+		}
 	}
 
-	onCopy() {
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.samplePermalinkData !== nextProps.samplePermalinkData ) {
+			this.updateSamplePermalink( nextProps.samplePermalinkData[ 0 ], nextProps.samplePermalinkData[ 1 ] );
+		}
+	}
+
+	/**
+	 * Updates the samplePermalink data in the state.
+	 *
+	 * @param {string} template Permalink structure to replace the tags in.
+	 * @param {string} postName The post slug to use in replacements.
+	 */
+	updateSamplePermalink( template, postName ) {
+		const regex = /%(?:postname|pagename)%/;
 		this.setState( {
-			showCopyConfirmation: true,
+			samplePermalink: template.replace( regex, postName ),
+			samplePermalinkPrefix: template.split( regex )[ 0 ],
+			samplePermalinkSuffix: template.split( regex )[ 1 ],
+			postName: postName,
 		} );
 	}
 
-	onFinishCopy() {
+	onSavePermalink() {
+		const postName = this.state.editedPostName.replace( /\s+/g, '-' );
+
+		this.props.editPost( {
+			slug: postName,
+			sample_permalink: [ this.props.samplePermalinkData[ 0 ], postName ],
+		} );
+
 		this.setState( {
-			showCopyConfirmation: false,
+			editingPermalink: false,
+			editedPostName: postName,
 		} );
 	}
 
 	render() {
-		const { isNew, link } = this.props;
-		if ( isNew || ! link ) {
+		const { isNew, previewLink, permalinkStructure } = this.props;
+		const {
+			iconClass,
+			samplePermalink,
+			samplePermalinkPrefix,
+			samplePermalinkSuffix,
+			postName,
+			editingPermalink,
+		} = this.state;
+
+		if ( isNew || ! previewLink ) {
 			return null;
 		}
 
 		return (
 			<div className="editor-post-permalink">
-				<Dashicon icon="admin-links" />
-				<span className="editor-post-permalink__label">{ __( 'Permalink:' ) }</span>
-				<Button className="editor-post-permalink__link" href={ link } target="_blank">
-					{ decodeURI( link ) }
-				</Button>
 				<ClipboardButton
-					className="button"
-					text={ link }
-					onCopy={ this.onCopy }
-					onFinishCopy={ this.onFinishCopy }
+					className="editor-post-permalink__copy"
+					text={ samplePermalink }
+					title={ __( 'Copy the permalink to your clipboard.' ) }
+					onCopy={ () => this.setState( { iconClass: 'copied' } ) }
 				>
-					{ this.state.showCopyConfirmation ? __( 'Copied!' ) : __( 'Copy' ) }
+					<Dashicon icon="admin-links" className={ iconClass } />
 				</ClipboardButton>
+
+				<span className="editor-post-permalink__label">{ __( 'Permalink:' ) }</span>
+
+				{ // Show the sample permalink when it isn't being editing.
```

Comment:
If we feel the need to comment inline in the render, it might be a sign there's enough complexity to warrant extracting out into a separate named/documented function. Or better yet, a separate component.

Reply:
Yah, it was getting a bit complex. Separated the form out.